shared-libs/cht-datasource/src/local/person.ts

```diff
@@ -82,4 +89,138 @@ export namespace v1 {
+      // Check whether parent doc's `contact_type` or `type`(if `contact_type` is absent)
+      // matches with any of the allowed parents type.
+       
+      const typeToMatch = (parentDoc as Input.v1.PersonInput).contact_type 
+      || (parentDoc as Input.v1.PersonInput).type;
+      const parentTypeMatchWithAllowedParents = (contactTypeObject.parents as string[])
+        .find(parent => parent === typeToMatch);
+
+      if (!(parentTypeMatchWithAllowedParents)) {
+        throw new InvalidArgumentError(
+          `Parent of type ${JSON.stringify(typeToMatch)} is not allowed for ${JSON.stringify(
+            input.contact_type
+          )} type`
+        );
+      }
+      return parentDoc;
+    };
+
+    const validatePersonParent = async (
+      contactTypeObject: Record<string, unknown>,
+      input: Record<string, unknown>
+    ): Promise<Nullable<Doc>> => {
+      if (hasField(contactTypeObject, {name: 'parents', type: 'object'})) {
+        return await ensureHasValidParentFieldAndReturnParentDoc(input, contactTypeObject);
+      }
+
+      throw new InvalidArgumentError(
+        `Invalid type of person, cannot have parent for ${
+          JSON.stringify(input.contact_type)
+        } type`
+      );
+    };
+
+    const appendParent = async (
+      typeFoundInSettingsContactTypes: Record<string, unknown> | undefined,
+      input: Input.v1.PersonInput
+    ) => {
+      let parentDoc: Nullable<Doc> = null;
+      if (typeFoundInSettingsContactTypes) {
+        parentDoc = await validatePersonParent(typeFoundInSettingsContactTypes, input);
+      } else if (input.parent) {
+        parentDoc = await getPersonDoc(input.parent);
+      }
+
+      if (parentDoc === null) {
+        throw new InvalidArgumentError(
+          `Parent with _id ${input.parent} does not exist.`
+        );
+      }
+      input = addParentToInput(input, 'parent', parentDoc);
+      return input;
+    };
+
+    return async (input: Input.v1.PersonInput): Promise<Person.v1.Person> => {
+      input = Input.v1.validatePersonInput(input);
+      if (hasField(input, { name: '_rev', type: 'string', ensureTruthyValue: true })) {
+        throw new InvalidArgumentError('Cannot pass `_rev` when creating a person.');
+      }
+
+      // This check can only be done when we have the contact_types from LocalDataContext.
+      const allowedContactTypes = getContactTypes(settings.getAll());
+      const typeFoundInSettingsContactTypes = allowedContactTypes.find(type => type.id === input.type);
+      const typeIsHardCodedPersonType = input.type === 'person';
+      if (!typeFoundInSettingsContactTypes && !typeIsHardCodedPersonType) {
+        throw new InvalidArgumentError('Invalid person type.');
+      }
+
+      // Append `contact_type` for newer versions.
+      if (typeFoundInSettingsContactTypes) {
+        input = {
+          ...input,
+          contact_type: input.type,
+          type: 'contact'
+        } as unknown as Input.v1.PersonInput;
+      }
```

Comment:
Minor but this can be simplified a bit. There is a dedicated function for getting person types:

```suggestion
      const customPersonType = contactTypeUtils
        .getPersonTypes(settings.getAll())
        .find(type => type.id === input.type);
      if (!customPersonType && input.type !== 'person') {
        throw new InvalidArgumentError('Invalid person type.');
      }
      if (customPersonType) {
        input = {
          ...input,
          contact_type: input.type,
          type: 'contact'
        };
      }
```